Hashtable.c

```diff
@@ -79,10 +135,42 @@ void Hashtable_delete(Hashtable* this) {
    free(this);
 }
 
-void Hashtable_put(Hashtable* this, unsigned int key, void* value) {
+void Hashtable_put(Hashtable* this, hkey_t key, void* value) {
+
+   assert(Hashtable_isConsistent(this));
+   assert(this->size > 0);
+
+   // grow on load > 0.7
+   if (10 * this->items + 1 > 7 * this->size) {
+      unsigned int newSize = nextPrime(2 * this->size);
+      HashtableItem** newBuckets = (HashtableItem**) xCalloc(newSize, sizeof(HashtableItem*));
+      // rehash
+      for (unsigned int i = 0; i < this->size; i++) {
+         HashtableItem* walk = this->buckets[i];
+         while (walk) {
+            HashtableItem* next = walk->next;
+            walk->next = NULL;
+
+            unsigned int newIndex = walk->key % newSize;
+            HashtableItem** newBucketPtr = &newBuckets[newIndex];
+            while (*newBucketPtr)
+               newBucketPtr = &(*newBucketPtr)->next;
+            *newBucketPtr = walk;
+
+            walk = next;
+         }
```

Comment:
Split resizing into its own routine? Useful if you know you are adding quite a few items soon.